---
title: "Final Project DSC 291"
author: "Eva Jason"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(qqman)
library(kableExtra)
library(RIdeogram)
```

## Inspect the results from the TCGA dataset

```{r, eval = F}
res_tcga_h = list.files(path="../fusion_twas-master/TCGA", pattern = "TCGA")
res_list_tcga_h = c()
for (i in 1:length(res_tcga_h)){
  b = read.table(paste0("../fusion_twas-master/TCGA/", res_tcga_h[[i]]), header=T) %>%
    drop_na(TWAS.P) %>%
    mutate(TWAS.P_adj_fdr  = p.adjust(TWAS.P, "fdr")) %>%
    filter(TWAS.P_adj_fdr <= 0.05) %>%
    mutate(dataset = "TCGA")
  res_list_tcga_h[[i]] = b
}
res_list_tcga_h = do.call(rbind, res_list_tcga_h)
```

## Inspect the results from TCGA SP dataset

```{r}
res_tcga_sp = list.files(path="../fusion_twas-master/TCGA_SP", pattern = "TCGA")
res_list_tcga_sp = c()
for (i in 1:length(res_tcga_sp)){
  b = read.table(paste0("../fusion_twas-master/TCGA_SP/", res_tcga_sp[[i]]), header=T) %>%
    drop_na(TWAS.P) %>%
    mutate(TWAS.P_adj_fdr  = p.adjust(TWAS.P, "fdr")) %>%
    filter(TWAS.P_adj_fdr <= 0.05)  %>%
    mutate(dataset = "TCGA_SP")
  res_list_tcga_sp[[i]] = b
}
res_list_tcga_sp = do.call(rbind, res_list_tcga_sp)
```

## Inspect the results from TCGA GE dataset

```{r, eval=F}
res_tcga_ge = list.files(path="../fusion_twas-master/TCGA_GE", pattern = "TCGA")
res_list_tcga_ge = c()
for (i in 1:length(res_tcga_ge)){
  b = read.table(paste0("../fusion_twas-master/TCGA_GE/", res_tcga_ge[[i]]), header=T) %>%
    drop_na(TWAS.P) %>%
    mutate(TWAS.P_adj_fdr  = p.adjust(TWAS.P, "fdr")) %>%
    filter(TWAS.P_adj_fdr <= 0.05)  %>%
    mutate(dataset = "TCGA_GE")
  res_list_tcga_ge[[i]] = b
}
res_list_tcga_ge = do.call(rbind, res_list_tcga_ge)
```

## Another dataset can be inspected as follows:

Replace the paths for the `.dat` files and the pattern specified in the bash 
file for the output. Change the 'dataset' variable for the name of the source of
the data

```{r, eval=F}
res = list.files(path="path/to/datfiles", pattern = "ABCD")
res_list = c()
for (i in 1:length(res)){
  b = read.table(paste0("path/to/datfiles", res_list[[i]]), header=T) %>%
    drop_na(TWAS.P) %>%
    mutate(TWAS.P_adj_fdr  = p.adjust(TWAS.P, "fdr")) %>%
    filter(TWAS.P_adj_fdr <= 0.05)  %>%
    mutate(dataset = "dataset")
  res_list[[i]] = b
}
res_list = do.call(rbind, res_list)
```

## Combine the datasets into one data frame

```{r, eval=F}
twas_all_results = rbind(res_list_tcga_h, res_list_tcga_sp, res_list_tcga_ge) %>%
  select(BEST.GWAS.ID, BEST.GWAS.Z, ID, TWAS.Z, TWAS.P_adj_fdr, dataset) %>%
  rename(SNP = BEST.GWAS.ID)
```

## Compare to GWAS

```{r}
twasResults = read.table("../GTEx_2ndGWAS_TWASsig_genes") %>%
  mutate(rsid = BEST.GWAS.ID)
```

```{r}
gwasResults = fread("../fusion_twas-master/GCST90454347.rename.h.tsv")
  #mutate(log10_pval = -log10(p_value))
```
Here we look at the SNPs that were found in the TWAS compared to those found in GWAS

```{r}
df_match = merge(gwasResults, twasResults)# %>%
  #select(SNP, ID, hm_beta, p_value, beta, TWAS.P_adj_fdr)

kable(df_match)
```

```{r}
qq(gwasResults$p_value)
```

Prepare the Manhattan Plot:

Code derived from https://r-graph-gallery.com/101_Manhattan_plot.html

```{r, cache=T}
don <- gwasResults %>% 
  
  # Compute chromosome size
  group_by(chromosome) %>% 
  summarise(chr_len=max(base_pair_location)) %>% 
  
  # Calculate cumulative position of each chromosome
  mutate(tot=cumsum(as.numeric(chr_len))-as.numeric(chr_len)) %>%
  select(-chr_len) %>%
  
  # Add this info to the initial dataset
  left_join(gwasResults, ., by=c("chromosome"="chromosome")) %>%
  
  # Add a cumulative position of each SNP
  arrange(chromosome, base_pair_location) %>%
  mutate( BPcum=base_pair_location+tot) %>%

  # Add highlight and annotation information
  mutate( is_annotate = ifelse(rsid %in% twasResults$rsid, "yes", "no"))# %>%
  #mutate( is_annotate= ifelse(rsid %in% twasResults$rsid, "yes", "no")) 

axisdf = don %>%
  
  group_by(chromosome) %>%
  summarize(center=(as.numeric(max(BPcum)) + as.numeric(min(BPcum) )) / 2 )
```


```{r}
cutoff = 5e-8
man_plot = ggplot(don, aes(x=BPcum, y=-log10(p_value))) +
    
    # Show all points
    geom_point( aes(color=as.factor(chromosome)), alpha=0.8, size=1.3) +
    scale_color_manual(values = rep(c("grey", "skyblue"), 22 )) +
    
    geom_hline(yintercept=-log10(cutoff), 
                color = "red", linewidth=1) +
    # custom X axis:
    scale_x_continuous( label = axisdf$chromosome, breaks= axisdf$center ) +
    scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis
    
    # Add label using ggrepel to avoid overlapping
    #geom_label_repel( data=subset(don, is_annotate=="yes"), aes(label=rsid), size=2, max.overlaps = 30) +
    
    
    # Add highlighted points
    #geom_point(data=subset(don, is_annotate=="yes"), color="orange", size=2) +
  
    # Custom the theme:
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )
man_plot

ggsave("manhattan_plot.png")

#man_plot + ylim(0,20)
ggsave("manhattan.png")
```

```{r, cache=T}
gwasResults15 = filter(gwasResults, neg_log_10_p_value < 16)
don15 <- gwasResults15 %>% 
  #filter(neg_log_10_p_value < 20) %>%
  
  # Compute chromosome size
  group_by(chromosome) %>% 
  summarise(chr_len=max(base_pair_location)) %>% 
  
  # Calculate cumulative position of each chromosome
  mutate(tot=cumsum(as.numeric(chr_len))-as.numeric(chr_len)) %>%
  select(-chr_len) %>%
  
  # Add this info to the initial dataset
  left_join(gwasResults15, ., by=c("chromosome"="chromosome")) %>%
  
  # Add a cumulative position of each SNP
  arrange(chromosome, base_pair_location) %>%
  mutate( BPcum=base_pair_location+tot) %>%

  # Add highlight and annotation information
  mutate( is_annotate = ifelse(rsid %in% twasResults$rsid, "yes", "no"))# %>%
  #mutate( is_annotate= ifelse(rsid %in% twasResults$rsid, "yes", "no")) 

axisdf15 = don15 %>%
  
  group_by(chromosome) %>%
  summarize(center=(as.numeric(max(BPcum)) + as.numeric(min(BPcum) )) / 2 )
```


```{r}
cutoff = 5e-8
man_plot15 = ggplot(don15, aes(x=BPcum, y=-log10(p_value))) +
    
    # Show all points
    geom_point( aes(color=as.factor(chromosome)), alpha=0.8, size=1.3) +
    scale_color_manual(values = rep(c("grey", "skyblue"), 22 )) +
    
    geom_hline(yintercept=-log10(cutoff), 
                color = "red", linewidth=1) +
    # custom X axis:
    scale_x_continuous( label = axisdf$chromosome, breaks= axisdf$center ) +
    scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis
    
    # Add label using ggrepel to avoid overlapping
    geom_label_repel( data=subset(don15, is_annotate=="yes"), aes(label=rsid), size=2, max.overlaps = 30) +
    
    
    # Add highlighted points
    geom_point(data=subset(don15, is_annotate=="yes"), color="orange", size=2) +
  
    # Custom the theme:
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )
man_plot20

ggsave("manhattan_plot15.png")

```

## Create Ideogram

Documentation can be found here:
https://cran.r-project.org/web/packages/RIdeogram/vignettes/RIdeogram.html

A gene density

```{r}
cutoff = 5e-8
gwasResults_sig = filter(gwasResults, p_value <= cutoff)
shared_sig = intersect(gwasResults_sig$rsid, twasResults$rsid)


twasIdeo = filter(twasResults, !rsid %in% shared_sig) %>%
  select(CHR, P0, P1) %>%
  rename(Chr = CHR,  Start = P0, End = P1) %>%
  mutate(Chr = as.integer(Chr), Start = as.integer(Start), End = as.numeric(End), Type = "TWAS", Shape = "box", color = "33a02c") #%>%
  #mutate(Type = ifelse(df_match$rsid %in% rsid, "Both", Type))

gwasIdeo = filter(gwasResults_sig, !rsid %in% shared_sig) %>%
  select(base_pair_location, chromosome) %>%
  mutate(End = as.integer(base_pair_location+1), 
         base_pair_location = as.integer(base_pair_location),
         Type = "GWAS", Shape = "circle", color = "6a3d9a") %>%
  rename(Start = base_pair_location, Chr = chromosome)

shared = filter(twasResults, rsid %in% shared_sig) %>%
  select(CHR, P0, P1) %>%
  rename(Chr = CHR,  Start = P0, End = P1) %>%
  mutate(Chr = as.integer(Chr), Start = as.integer(Start), End = as.numeric(End), Type = "Both", Shape = "triangle", color = "ff7f00")

dfIdeo = rbind(twasIdeo, gwasIdeo) %>%
  rbind(shared) %>%
  select(Type, Shape, Chr, Start, End, color)

ideogram(karyotype = human_karyotype, overlaid = gene_density, label = dfIdeo, label_type = "marker")
convertSVG("chromosome.svg", device = "png", dpi=600)
```

```{r}
df_match_table = select(df_match, rsid, Z, p_value, TWAS.Z, FDR, ID) %>%
  mutate(rsid_id = paste0(rsid, "/", ID)) %>%
  select(-rsid, -ID) %>%
  mutate(p_value = format(p_value, scientific = T), p_value_adj =  format(FDR, scientific = T))
df_match_table_gwas = select(df_match_table, rsid_id, Z, p_value) %>%
  mutate(study = "GWAS") %>%
  mutate(significant = ifelse(as.numeric(p_value) < cutoff, "*", ""))
df_match_table_twas = select(df_match_table, rsid_id, TWAS.Z, p_value_adj, FDR) %>%
  mutate(significant = ifelse(FDR < 0.05,"*", "")) %>%
  rename(Z = TWAS.Z, p_value = p_value_adj) %>%
  mutate(study = "TWAS") %>%
  select(-FDR)

df_match_long = rbind(df_match_table_gwas, df_match_table_twas)

bar_match = ggplot(df_match_long, aes(fill=study, y=Z, x=rsid_id)) + 
  geom_bar(position="dodge", stat="identity") +
  geom_text(aes(label=significant), position=position_dodge(width=0.9)) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5), axis.title.y=element_blank()) +
  coord_flip()
bar_match
ggsave("bar_match.png")

write.table(df_match_long, "table_matching_twas_gwas.tsv", row.names = F, quote = F, sep = "\t")

df_match_wide = select(df_match, rsid, Z, p_value, ID, TWAS.Z, FDR) %>%
  mutate(Z = round(Z, 2),TWAS.Z = round(TWAS.Z, 2)) %>%
  mutate(p_value = format(p_value, scientific = T), p_value_adj =  format(FDR, scientific = T))
write.table(df_match_wide, "table_matching_twas_gwas_wide.tsv", row.names = F, quote = F, sep = "\t")
```

```{r}
only_sig = filter(df_match, FDR < 0.05, p_value < cutoff) %>%
  select(rsid, chromosome, base_pair_location, ID, CHR, P0,P1, FDR, p_value, Z, TWAS.Z)
write.table(only_sig, "onlysig.tsv", sep = "\t", quote = F, row.names = F)
```




```{r}
write.table(gwasResults_sig$rsid, "gwasResults_sig_rsid.txt", sep = "\t", quote = F, row.names = F, col.names = F)
write.table(twasResults$ID, "twasResults_sig_ID.txt", sep = "\t", quote = F, row.names = F, col.names = F)
```



