---
title: "Final Project DSC 291"
author: "Eva Jason"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(qqman)
library(kableExtra)
library(RIdeogram)
```

## Read TWAS Result Files

If more TWASes are preformed, they can be added here.

```{r}
twasResults = read.table("GTEx_2ndGWAS_TWASsig_genes") %>%
  mutate(rsid = BEST.GWAS.ID, model = "GTEx")

twasResults2 = read.table("breastcancer_tcga_tumor_ex2.top") %>%
  mutate(rsid = BEST.GWAS.ID, model = "TCGA")

twasResults = rbind(twasResults, twasResults2)
```

## Read GWAS summary statistics

```{r}
# Genome wide significance cutoff:
cutoff = 5e-8

gwasResults = fread("../fusion_twas-master/GCST90454347.rename.h.tsv")
qq(gwasResults$p_value)
```

Prepare the Manhattan Plot:

Code derived from https://r-graph-gallery.com/101_Manhattan_plot.html

```{r, cache=T}
don  = gwasResults %>% 
  
  # Compute chromosome size
  group_by(chromosome) %>% 
  summarise(chr_len=max(base_pair_location)) %>% 
  
  # Calculate cumulative position of each chromosome
  mutate(tot=cumsum(as.numeric(chr_len))-as.numeric(chr_len)) %>%
  select(-chr_len) %>%
  
  # Add this info to the initial dataset
  left_join(gwasResults, ., by=c("chromosome"="chromosome")) %>%
  
  # Add a cumulative position of each SNP
  arrange(chromosome, base_pair_location) %>%
  mutate( BPcum=base_pair_location+tot)

axisdf = don %>%
  group_by(chromosome) %>%
  summarize(center=(as.numeric(max(BPcum)) + as.numeric(min(BPcum) )) / 2 )

man_plot = ggplot(don, aes(x=BPcum, y=-log10(p_value))) +
    
    # Show all points
    geom_point( aes(color=as.factor(chromosome)), alpha=0.8, size=1.3) +
    scale_color_manual(values = rep(c("grey", "skyblue"), 22 )) +
    
    geom_hline(yintercept=-log10(cutoff), 
                color = "red", linewidth=1) +
    # custom X axis:
    scale_x_continuous( label = axisdf$chromosome, breaks= axisdf$center ) +
    scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis

    # Custom the theme:
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )
man_plot

ggsave("manhattan_plot.png")
```

## Create Ideogram

Documentation can be found here:
https://cran.r-project.org/web/packages/RIdeogram/vignettes/RIdeogram.html

A gene density overlay is used.

```{r}
gwasResults_sig = filter(gwasResults, p_value <= cutoff)
shared_sig = intersect(gwasResults_sig$rsid, twasResults$rsid)

twasIdeo = filter(twasResults, !rsid %in% shared_sig) %>%
  select(CHR, P0, P1) %>%
  rename(Chr = CHR,  Start = P0, End = P1) %>%
  mutate(Chr = as.integer(Chr), Start = as.integer(Start), 
         End = as.numeric(End), Type = "TWAS", Shape = "box", color = "33a02c")

gwasIdeo = filter(gwasResults_sig, !rsid %in% shared_sig) %>%
  select(base_pair_location, chromosome) %>%
  mutate(End = as.integer(base_pair_location+1), 
         base_pair_location = as.integer(base_pair_location),
         Type = "GWAS", Shape = "circle", color = "6a3d9a") %>%
  rename(Start = base_pair_location, Chr = chromosome)

sharedIdeo = filter(twasResults, rsid %in% shared_sig) %>%
  select(CHR, P0, P1) %>%
  rename(Chr = CHR,  Start = P0, End = P1) %>%
  mutate(Chr = as.integer(Chr), Start = as.integer(Start), End = as.numeric(End), Type = "Both", Shape = "triangle", color = "ff7f00")

dfIdeo = rbind(twasIdeo, gwasIdeo) %>%
  rbind(sharedIdeo) %>%
  select(Type, Shape, Chr, Start, End, color)

ideogram(karyotype = human_karyotype, overlaid = gene_density, label = dfIdeo, label_type = "marker")

# increased quality to 600
convertSVG("chromosome.svg", device = "png", dpi=600)
```

## Merge TWAS and GWAS results

Get a data frame that has the full set of results. The merge is performed by rsID. 
The rsID for the twas corresponds to the columns `BEST.GWAS.ID`. 

```{r}
match_df = merge(twasResults, gwasResults) %>%
  # remove rows where the GWAS p-value is less than the cutoff
  filter(p_value <= cutoff) %>%
  mutate(rsid_id = paste0(rsid, "/", ID))
```

```{r}
match_df_gwas = select(match_df, p_value, rsid_id, Z) %>%
  mutate(model = "GWAS")

match_df_twas = select(match_df, FDR, rsid_id, TWAS.Z, model) %>%
  rename(p_value = FDR, Z = TWAS.Z)

df_match_long = rbind(match_df_gwas, match_df_twas)
write.table(df_match_long, "df_match_long.tsv",sep = "\t", quote = F, 
            row.names = F)

df_match_wide = cbind(select(match_df, p_value, rsid_id, Z, chromosome, 
                             base_pair_location), select(match_df, FDR, rsid_id, TWAS.Z, model))
write.table(df_match_wide, "df_match_wide.tsv", , sep = "\t", quote = F, 
            row.names = F)
```

This bar charts compares the affect sizes (Z-scores) of the variants from the 
TWAS and GWAS. Additionally the data set for the eQTL model in the TWAS is 
separated by color. 

```{r}
bar_match = ggplot(df_match_long, aes(fill=model, y=Z, x=rsid_id)) + 
  geom_bar(position="dodge", stat="identity") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5), 
        axis.title.y=element_blank(), legend.title=element_blank()) +
  coord_flip()
bar_match
ggsave("bar_match.png")
```


