---
title: "Final Project DSC 291"
author: "Eva Jason"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(dplyr)
library(tidyr)
library(ggplot2)
library(qqman)
library(kableExtra)
```

## Inspect the results from the TCGA dataset

```{r}
res_tcga_h = list.files(path="../fusion_twas-master/TCGA", pattern = "TCGA")
res_list_tcga_h = c()
for (i in 1:length(res_tcga_h)){
  b = read.table(paste0("../fusion_twas-master/TCGA/", res_tcga_h[[i]]), header=T) %>%
    drop_na(TWAS.P) %>%
    mutate(TWAS.P_adj_fdr  = p.adjust(TWAS.P, "fdr")) %>%
    filter(TWAS.P_adj_fdr <= 0.05) %>%
    mutate(dataset = "TCGA")
  res_list_tcga_h[[i]] = b
}
res_list_tcga_h = do.call(rbind, res_list_tcga_h)
```

## Inspect the results from TCGA SP dataset

```{r}
res_tcga_sp = list.files(path="../fusion_twas-master/TCGA_SP", pattern = "TCGA")
res_list_tcga_sp = c()
for (i in 1:length(res_tcga_sp)){
  b = read.table(paste0("../fusion_twas-master/TCGA_SP/", res_tcga_sp[[i]]), header=T) %>%
    drop_na(TWAS.P) %>%
    mutate(TWAS.P_adj_fdr  = p.adjust(TWAS.P, "fdr")) %>%
    filter(TWAS.P_adj_fdr <= 0.05)  %>%
    mutate(dataset = "TCGA_SP")
  res_list_tcga_sp[[i]] = b
}
res_list_tcga_sp = do.call(rbind, res_list_tcga_sp)
```

## Inspect the results from TCGA GE dataset

```{r}
res_tcga_ge = list.files(path="../fusion_twas-master/TCGA_GE", pattern = "TCGA")
res_list_tcga_ge = c()
for (i in 1:length(res_tcga_ge)){
  b = read.table(paste0("../fusion_twas-master/TCGA_GE/", res_tcga_ge[[i]]), header=T) %>%
    drop_na(TWAS.P) %>%
    mutate(TWAS.P_adj_fdr  = p.adjust(TWAS.P, "fdr")) %>%
    filter(TWAS.P_adj_fdr <= 0.05)  %>%
    mutate(dataset = "TCGA_GE")
  res_list_tcga_ge[[i]] = b
}
res_list_tcga_ge = do.call(rbind, res_list_tcga_ge)
```

## Another dataset can be inspected as follows:

Replace the paths for the `.dat` files and the pattern specified in the bash 
file for the output. Change the 'dataset' variable for the name of the source of
the data

```{r, eval=F}
res = list.files(path="path/to/datfiles", pattern = "ABCD")
res_list = c()
for (i in 1:length(res)){
  b = read.table(paste0("path/to/datfiles", res_list[[i]]), header=T) %>%
    drop_na(TWAS.P) %>%
    mutate(TWAS.P_adj_fdr  = p.adjust(TWAS.P, "fdr")) %>%
    filter(TWAS.P_adj_fdr <= 0.05)  %>%
    mutate(dataset = "dataset")
  res_list[[i]] = b
}
res_list = do.call(rbind, res_list)
```

## Combine the datasets into one data frame

```{r}
twas_all_results = rbind(res_list_tcga_h, res_list_tcga_sp, res_list_tcga_ge) %>%
  select(BEST.GWAS.ID, BEST.GWAS.Z, ID, TWAS.Z, TWAS.P_adj_fdr, dataset) %>%
  rename(SNP = BEST.GWAS.ID)
```

## Compare to GWAS

Read in the same GWAS used in the TWAS study.

```{r}
gwasResults = fread("../fusion_twas-master/34737426-GCST90043925-EFO_1000326.rename.h.tsv") %>%
  mutate(log10_pval = -log10(p_value))
```
Here we look at the SNPs that were found in the TWAS compared to those found in GWAS

```{r}
#gwasResults_twas_match = filter(gwasResults, SNP %in% twas_all_results$SNP)
df_match = merge(gwasResults_twas_match, twas_all_results, by = "SNP") %>%
  select(SNP, ID, hm_beta, p_value, beta, TWAS.P_adj_fdr)

kable(df_match)
```

```{r}
cutoff = 5e-8
gwasResults_sig = filter(gwasResults, p_value <= cutoff)
```

```{r}
qq(gwasResults$p_value)
```

Prepare the Manhattan Plot:

Code derived from https://r-graph-gallery.com/101_Manhattan_plot.html

```{r, cache=T}
don <- gwasResults %>% 
  
  # Compute chromosome size
  group_by(chromosome) %>% 
  summarise(chr_len=max(base_pair_location)) %>% 
  
  # Calculate cumulative position of each chromosome
  mutate(tot=cumsum(as.numeric(chr_len)-as.numeric(chr_len))) %>%
  select(-chr_len) %>%
  
  # Add this info to the initial dataset
  left_join(gwasResults, ., by=c("chromosome"="chromosome")) %>%
  
  # Add a cumulative position of each SNP
  arrange(chromosome, base_pair_location) %>%
  mutate( BPcum=base_pair_location+tot)

axisdf = don %>%
  
  group_by(chromosome) %>%
  summarize(center=((as.numeric(max(BPcum)) + as.numeric(min(BPcum) ))) / 2 )
```


```{r}
man_plot = ggplot(don, aes(x=BPcum, y=-log10(p_value))) +
    
    # Show all points
    geom_point( aes(color=as.factor(chromosome)), alpha=0.8, size=1.3) +
    scale_color_manual(values = rep(c("grey", "skyblue"), 22 )) +
    
    geom_hline(yintercept=-log10(cutoff), 
                color = "red", linewidth=1) +
    # custom X axis:
    scale_x_continuous( label = axisdf$chromosome, breaks= axisdf$center ) +
    scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis
  
    # Custom the theme:
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )
man_plot

ggsave("manhattan_plot.png")
```




